File: kobra_app/lib/services/medios_pago_service.dart
import 'api_client.dart';
import '../models/medio_pago.dart';

class MediosPagoService {
  Future<List<MedioPago>> listar() async {
    final response = await ApiClient.get<List<dynamic>>('/medios-pago');
    return response.data!.map((e) => MedioPago.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<MedioPago> crear(String nombre) async {
    final response = await ApiClient.post<Map<String, dynamic>>(
      '/medios-pago',
      data: {'nombre': nombre},
    );
    return MedioPago.fromJson(response.data!);
  }

  Future<MedioPago> toggleActivo(int id) async {
    final response = await ApiClient.patch<Map<String, dynamic>>('/medios-pago/$id/toggle');
    return MedioPago.fromJson(response.data!);
  }

  Future<void> eliminar(int id) async {
    await ApiClient.delete('/medios-pago/$id');
  }
}

File: kobra-backend/src/medios-pago/medios-pago.service.ts
import { BadRequestException, Injectable, NotFoundException } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { UsuarioActual } from '../common/decorators/current-user.decorator';
import { CreateMedioPagoDto } from './dto/create-medio-pago.dto';

const DEFAULTS = ['Efectivo', 'Transferencia', 'Tarjeta'];
const ORDER = [{ esDefault: 'desc' as const }, { nombre: 'asc' as const }];

@Injectable()
export class MediosPagoService {
  constructor(private prisma: PrismaService) {}

  async findAll(usuario: UsuarioActual) {
    const existing = await this.prisma.medioPago.findMany({
      where: { negocioId: usuario.negocioId },
      orderBy: ORDER,
    });

    if (existing.length === 0) {
      await this.prisma.medioPago.createMany({
        data: DEFAULTS.map((nombre) => ({
          negocioId: usuario.negocioId,
          nombre,
          esDefault: true,
          activo: true,
        })),
      });
      return this.prisma.medioPago.findMany({
        where: { negocioId: usuario.negocioId },
        orderBy: ORDER,
      });
    }

    return existing;
  }

  async create(dto: CreateMedioPagoDto, usuario: UsuarioActual) {
    return this.prisma.medioPago.create({
      data: {
        negocioId: usuario.negocioId,
        nombre: dto.nombre,
        activo: true,
        esDefault: false,
      },
    });
  }

  async toggleActivo(id: number, usuario: UsuarioActual) {
    const medio = await this.prisma.medioPago.findFirst({
      where: { id, negocioId: usuario.negocioId },
    });
    if (!medio) throw new NotFoundException(`Medio de pago ${id} no encontrado`);
    return this.prisma.medioPago.update({
      where: { id },
      data: { activo: !medio.activo },
    });
  }

  async remove(id: number, usuario: UsuarioActual) {
    const medio = await this.prisma.medioPago.findFirst({
      where: { id, negocioId: usuario.negocioId },
    });
    if (!medio) throw new NotFoundException(`Medio de pago ${id} no encontrado`);
    if (medio.esDefault) {
      throw new BadRequestException('Los medios predeterminados no se pueden eliminar, solo deshabilitar');
    }
    return this.prisma.medioPago.delete({ where: { id } });
  }
}
